src/parser/base.rs

```diff
@@ -0,0 +1,283 @@
+
+use crate::field::LurkField;
+use base_x;
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, take_till},
+    character::complete::satisfy,
+    combinator::value,
+    error::context,
+    IResult, InputTakeAtPosition,
+};
+
+use crate::parser::{
+    error::{map_parse_err, ParseError, ParseErrorKind},
+    Span,
+};
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum LitBase {
+    Bin,
+    Oct,
+    Dec,
+    Hex,
+}
+
+impl Default for LitBase {
+    fn default() -> Self {
+        Self::Hex
+    }
+}
+
+impl LitBase {
+    pub fn parse_code<F: LurkField>(
+        i: Span<'_>,
+    ) -> IResult<Span<'_>, Self, ParseError<Span<'_>, F>> {
+        alt((
+            value(Self::Bin, tag("b")),
+            value(Self::Oct, tag("o")),
+            value(Self::Dec, tag("d")),
+            value(Self::Hex, tag("x")),
+        ))(i)
+    }
+
+    /// Get the code corresponding to the base algorithm.
+    pub fn code(&self) -> char {
+        match self {
+            Self::Bin => 'b',
+            Self::Oct => 'o',
+            Self::Dec => 'd',
+            Self::Hex => 'x',
+        }
+    }
+
+    pub fn base_digits(&self) -> &str {
+        match self {
+            Self::Bin => "01",
+            Self::Oct => "01234567",
+            Self::Dec => "0123456789",
+            Self::Hex => "0123456789abcdef",
+        }
+    }
+
+    pub fn radix(&self) -> u32 {
+        match self {
+            Self::Bin => 2,
+            Self::Oct => 8,
+            Self::Dec => 10,
+            Self::Hex => 16,
+        }
+    }
+
+    pub fn is_digit(&self, x: char) -> bool {
+        self.base_digits()
+            .chars()
+            .any(|y| x.to_ascii_lowercase() == y)
+    }
+
+    pub fn encode<I: AsRef<[u8]>>(&self, input: I) -> String {
+        base_x::encode(self.base_digits(), input.as_ref())
+    }
+
+    pub fn decode<'a, F: LurkField>(
+        &self,
+        input: Span<'a>,
+    ) -> IResult<Span<'a>, Vec<u8>, ParseError<Span<'a>, F>> {
+        let (i, o) = input.split_at_position_complete(|x| !self.is_digit(x))?;
+        match base_x::decode(self.base_digits(), o.fragment()) {
+            Ok(bytes) => Ok((i, bytes)),
+            Err(_) => Err(nom::Err::Error(ParseError::new(
+                i,
+                ParseErrorKind::InvalidBaseEncoding(*self),
+            ))),
+        }
+    }
+
+    pub fn decode1<'a, F: LurkField>(
+        &self,
+        input: Span<'a>,
+    ) -> IResult<Span<'a>, Vec<u8>, ParseError<Span<'a>, F>> {
+        let (i, o) = input
+            .split_at_position1_complete(|x| !self.is_digit(x), nom::error::ErrorKind::Digit)?;
+        match base_x::decode(self.base_digits(), o.fragment()) {
+            Ok(bytes) => Ok((i, bytes)),
+            Err(_) => Err(nom::Err::Error(ParseError::new(
+                i,
+                ParseErrorKind::InvalidBaseEncoding(*self),
+            ))),
+        }
+    }
+}
+
+pub fn parse_bin_digits<F: LurkField>(
```

Comment:
This might help
```rust
macro_rules! define_parse_digits {
    ($name:ident, $base:ident, $digit_str:expr, $digits_str:expr, $map_fn:expr) => {
        pub fn $name<F: LurkField>() -> impl Fn(Span<'_>) -> IResult<Span<'_>, String, ParseError<Span<'_>, F>> {
            move |from: Span<'_>| {
                let (i, d) = context($digit_str, satisfy(|x| LitBase::$base.is_digit(x)))(from)?;
                let (i, ds) = context(
                    $digits_str,
                    take_till(|x| !(LitBase::$base.is_digit(x) || x == '_')),
                )(i)?;
                let ds: String = core::iter::once(d)
                    .chain((*ds.fragment()).to_owned().chars())
                    .filter(|x| *x != '_')
                    .map($map_fn)
                    .collect();
                Ok((i, ds))
            }
        }
    };
}

define_parse_digits!(parse_bin_digits, Bin, "binary digit", "binary digits", |x| x);
define_parse_digits!(parse_oct_digits, Oct, "octal digit", "octal digits", |x| x);
define_parse_digits!(parse_dec_digits, Dec, "decimal digit", "decimal digits", |x| x);
define_parse_digits!(parse_hex_digits, Hex, "hexadecimal digit", "hexadecimal digits", |x| x.to_ascii_lowercase());
```